File: POOPdz/POOPdz/XMLFormatter.cpp
#include"XMLFormatter.h"
#include <regex>

Operation* XMLh::importComposit(std::string file, Image *i) {
	ifstream pic;
	pic.open(file, ios::in);
	std::string file_as_string;
	std::regex oper("<([0-9A-za-z. ]+) Konstanta=\"([0-9A-Za-z]+)\"/>");
	std::smatch n;
	int c;
	vector<Operation*> temp;
	Operation* o;
	while (!regex_search(file_as_string, n, oper) && file_as_string != "</Composit>") {
		std::getline(pic, file_as_string);
	}
	while (regex_search(file_as_string, n, oper)) {
		bool flag = false;
		std::string name = n.str(1);
		if (n.str(2) != "nema")
			c = stoi(n.str(2));
		if (name == "Add") {
			o = i->takeOperation(name);
			o->setConstant(c);
			flag = true;
		}
		if (name == "Subtraction") {
			o = i->takeOperation(name);
			o->setConstant(c);
			flag = true;
		}
		if (name == "InvertedSubtraction") {
			o = i->takeOperation(name);
			o->setConstant(c);
			flag = true;
		}
		if (name == "Multiplication") {
			o = i->takeOperation(name);
			o->setConstant(c);
			flag = true;
		}
		if (name == "Divide") {
			o = i->takeOperation(name);
			o->setConstant(c);
			flag = true;
		}
		if (name == "InvertedDivision") {
			o = i->takeOperation(name);
			o->setConstant(c);
			flag = true;
		}
		if (name == "Power") {
			o = i->takeOperation(name);
			o->setConstant(c);
			flag = true;
		}
		if (name == "Log") {
			o = i->takeOperation(name);
			o->setConstant(c);
			flag = true;
		}
		if (name == "Absolute") {
			o = i->takeOperation(name);
			flag = true;
		}
		if (name == "Minimum") {
			o = i->takeOperation(name);
			o->setConstant(c);
			flag = true;
		}
		if (name == "Maximum") {
			o = i->takeOperation(name);
			o->setConstant(c);
			flag = true;
		}
		if (name == "Inversion") {
			o = i->takeOperation(name);
			flag = true;
		}
		if (name == "Grey") {
			o = i->takeOperation(name);
			flag = true;
		}
		if (name == "BlackAndWhite") {
			o = i->takeOperation(name);
			flag = true;
		}
		if (name == "Median") {
			o = i->takeOperation(name);
			flag = true;
		}
		if(!flag) {
			o = importComposit(name, i);
			temp.push_back(o);
			std::getline(pic, file_as_string);
			continue;
		}
		temp.push_back(o);
		std::getline(pic, file_as_string);
	}
	Layer help(1, 1);
	Operation* composit = new CompositOperation(temp, &help, file);
	i->addToCompositVector((CompositOperation*)composit);
	return composit;
}

Image XMLh::import(std::string file, Image *i) {
	Formatter* f1 = new BMPh;
	Formatter* f2 = new PAMh;
	ifstream pic;
	pic.open(file);
	std::string file_as_string;
	std::regex r("<Width width=\"([0-9]+)\"/>");
	std::smatch w;
	std::getline(pic, file_as_string);
	while (!regex_search(file_as_string, w, r)) { 
		std::getline(pic, file_as_string);
	}
	std::string wid = w.str(1);
	int width = stoi(wid);
	std::regex r1("<Height height=\"([0-9]+)\"/>");
	std::smatch h;
	while (!regex_search(file_as_string, h, r1)) {
		std::getline(pic, file_as_string);
	}
	std::string hh = h.str(1);
	int height = stoi(hh);
	std::regex lay("<Layer active=\"([a-z]+)\" name=\"([0-9A-Za-z\.]+)\" selected=\"([0-9]+)\" transparency=\"([0-9]+)\" visible=\"([a-z]+)\"/>");
	std::smatch l;
	while (!regex_search(file_as_string, h, lay)) {
		std::getline(pic, file_as_string);
	}
	while (regex_search(file_as_string, l, lay)) {
		bool visible = false, active = false;
		if (l.str(1) == "true")active = true;
		if (l.str(5) == "true")visible = true;
		int selected = stoi(l.str(3));
		int transparency = stoi(l.str(4));
		std::regex n("[0-9A-Za-z]+(.bmp)");
		std::smatch ex;
		std::string name = l.str(2);
		regex_search(name, ex, n);
		if (".bmp" == ex.str(1)) {
			*i = f1->import(name, i);
		}
		else
			*i = f2->import(name, i);
		for (Layer &l : i->getImage()) {
			if (l.getName() == name) {
				l.setActivity(active);
				l.setTransparency(transparency);
				l.setVisibility(visible);
				l.export_set_Selection(selected);
			}
		}
		std::getline(pic, file_as_string);
	}
	std::regex s("<Selection active=\"([A-Za-z]+)\" name=\"([0-9A-Za-z]+)\">");
	std::regex rec("<Rectangle height=\"([0-9]+)\" width=\"([0-9]+)\" x=\"([0-9]+)\" y=\"([0-9]+)\"/>");
	std::regex check("(</Selection>)");
	std::smatch ss;
	std::smatch	rr;
	std::smatch ch;
	while (!regex_search(file_as_string, ss, s) && file_as_string!="</Image>") {
		std::getline(pic, file_as_string);
	}
	while (regex_search(file_as_string, ss, s)) {
		std::string name = ss.str(2);
		bool active = false;
		if (ss.str(1) == "true")active = true;
		Selection select(name, active);
		std::getline(pic, file_as_string);
		while (regex_search(file_as_string, rr, rec)) {
			select.AddRectangle(stoi(rr.str(3)), stoi(rr.str(4)), stoi(rr.str(1)), stoi(rr.str(2)));
			i->addSelectionToMap(select.getName(), select);
			i->select(select.getName());
			std::getline(pic, file_as_string);
		}
		std::getline(pic, file_as_string);
	}
	std::regex comp("<Composit name=\"([0-9A-Za-z\.]+)\"/>");
	std::smatch mcomp;
	while (!regex_search(file_as_string, mcomp, comp) && file_as_string!="</Image>") {
		std::getline(pic, file_as_string);
	}
	while (regex_search(file_as_string, mcomp, comp)) {
		std::string name = mcomp.str(1);
		//std::string path = mcomp.str(2);
		importComposit(name, i);
		std::getline(pic, file_as_string);
	}
	return *i;
}
void XMLh::expor(Image& i, std::string file) {
	XMLDocument* xmlDoc = new XMLDocument();
	XMLNode * pRoot = xmlDoc->NewElement("Image");
	xmlDoc->InsertFirstChild(pRoot);
	XMLElement* pDimensions = xmlDoc->NewElement("Dimensions");
	XMLElement* pElement = xmlDoc->NewElement("Layers");
	Formatter *f = new BMPh();
	XMLElement* width = xmlDoc->NewElement("Width");
	int w = i.getImage().front().getLayer()[0].size();
	width->SetAttribute("width", w);
	pDimensions->InsertEndChild(width);
	XMLElement* height = xmlDoc->NewElement("Height");
	int h = i.getImage().front().getLayer().size();
	height->SetAttribute("height", h);
	pDimensions->InsertEndChild(height);
	pRoot->InsertEndChild(pDimensions);
	Image* temp = new Image();
	for (Layer l : i.getImage()) {	
		temp->addLayer(l);
		f->expor(*temp, l.getName());
		temp->deleteLayer(l);
		XMLElement* ElemLayer = xmlDoc->NewElement("Layer");
		ElemLayer->SetAttribute("active", l.getActivity());
		ElemLayer->SetAttribute("name", l.getName().c_str());
		ElemLayer->SetAttribute("selected", l.getSelection());
		ElemLayer->SetAttribute("transparency", l.getTransparency());
		ElemLayer->SetAttribute("visible", l.getVisibility());
		pElement->InsertEndChild(ElemLayer);		
	}
	pRoot->InsertEndChild(pElement);
	XMLElement* selections = xmlDoc->NewElement("Selections");
	for (pair<const std::string, Selection> s : i.get_Selection_Map()) {
		XMLElement* Selection = xmlDoc->NewElement("Selection");
		Selection->SetAttribute("active", s.second.getActivity());
		Selection->SetAttribute("name", s.first.c_str());
		for (Rectangle r : s.second.getRectrangles()) {
			XMLElement* SelectionRectangle = xmlDoc->NewElement("Rectangle");
			SelectionRectangle->SetAttribute("height", r.getHeight());
			SelectionRectangle->SetAttribute("width", r.getWidth());
			SelectionRectangle->SetAttribute("x", r.getX());
			SelectionRectangle->SetAttribute("y", r.getY());
			Selection->InsertEndChild(SelectionRectangle);
		}
		selections->InsertEndChild(Selection);
	}
	pRoot->InsertEndChild(selections);
	//XMLElement* Compelement = xmlDoc->NewElement("CompositFunctions");
	for (CompositOperation* j : i.getComposit_vector()) {
		exportComposit(i, j);
		//XMLElement* composit = xmlDoc->NewElement("Composit");
		//composit->SetAttribute("name", (j->getName()).c_str());
		//composit->SetAttribute("path", j->getPath().c_str());
		//Compelement->InsertEndChild(composit);
	}
	//pRoot->InsertEndChild(Compelement);
	xmlDoc->SaveFile(file.c_str());
}

void XMLh::exportComposit(Image i, CompositOperation* comp) {
	XMLDocument* xmlOperation = new XMLDocument();
	XMLNode* root = xmlOperation->NewElement("Composit");
	xmlOperation->InsertFirstChild(root);
	std::string path;
	std::string name = comp->getName().c_str();
	XMLElement* composit = xmlOperation->NewElement((comp->getName()).c_str());
	root->InsertEndChild(composit);
	path = "Composit/" + comp->getName();
	comp->setPath(path);
	//path = comp->getName() + ".fun";
	for (Operation* o : comp->getVector()) {
		if (typeid(*o) == typeid(Add)) {
			XMLElement* operation = xmlOperation->NewElement("Add");
			int c = o->getC();
			operation->SetAttribute("Konstanta", c);
			composit->InsertEndChild(operation);
		}
		if (typeid(*o) == typeid(Subtraction)) {
			XMLElement* operation = xmlOperation->NewElement("Subtraction");
			int c = o->getC();
			operation->SetAttribute("Konstanta", c);
			composit->InsertEndChild(operation);
		}
		if (typeid(*o) == typeid(InvSubtraction)) {
			XMLElement* operation = xmlOperation->NewElement("InvertedSubtraction");
			int c = o->getC();
			operation->SetAttribute("Konstanta", c);
			composit->InsertEndChild(operation);
		}
		if (typeid(*o) == typeid(Multiplication)) {
			XMLElement* operation = xmlOperation->NewElement("Multiplication");
			int c = o->getC();
			operation->SetAttribute("Konstanta", c);
			composit->InsertEndChild(operation);
		}
		if (typeid(*o) == typeid(Divide)) {
			XMLElement* operation = xmlOperation->NewElement("Divide");
			int c = o->getC();
			operation->SetAttribute("Konstanta", c);
			composit->InsertEndChild(operation);
		}
		if (typeid(*o) == typeid(InvDivide)) {
			XMLElement* operation = xmlOperation->NewElement("InvertedDivision");
			int c = o->getC();
			operation->SetAttribute("Konstanta", c);
			composit->InsertEndChild(operation);
		}
		if (typeid(*o) == typeid(Power)) {
			XMLElement* operation = xmlOperation->NewElement("Power");
			int c = o->getC();
			operation->SetAttribute("Konstanta", c);
			composit->InsertEndChild(operation);
		}
		if (typeid(*o) == typeid(Log)) {
			XMLElement* operation = xmlOperation->NewElement("Log");
			int c = o->getC();
			operation->SetAttribute("Konstanta", c);
			composit->InsertEndChild(operation);
		}
		if (typeid(*o) == typeid(Abs)) {
			XMLElement* operation = xmlOperation->NewElement("Absolute");
			operation->SetAttribute("Konstanta", "nema");
			composit->InsertEndChild(operation);
		}
		if (typeid(*o) == typeid(Min)) {
			XMLElement* operation = xmlOperation->NewElement("Minimum");
			int c = o->getC();
			operation->SetAttribute("Konstanta", c);
			composit->InsertEndChild(operation);
		}
		if (typeid(*o) == typeid(Max)) {
			XMLElement* operation = xmlOperation->NewElement("Maximum");
			int c = o->getC();
			operation->SetAttribute("Konstanta", c);
			composit->InsertEndChild(operation);
		}
		if (typeid(*o) == typeid(Inversion)) {
			XMLElement* operation = xmlOperation->NewElement("Inversion");
			operation->SetAttribute("Konstanta", "nema");
			composit->InsertEndChild(operation);
		}
		if (typeid(*o) == typeid(Grey)) {
			XMLElement* operation = xmlOperation->NewElement("Grey");
			operation->SetAttribute("Konstanta", "nema");
			composit->InsertEndChild(operation);
		}
		if (typeid(*o) == typeid(Black_and_white)) {
			XMLElement* operation = xmlOperation->NewElement("BlackAndWhite");
			operation->SetAttribute("Konstanta", "nema");
			composit->InsertEndChild(operation);
		}
		if (typeid(*o) == typeid(Median)) {
			XMLElement* operation = xmlOperation->NewElement("Medina");
			operation->SetAttribute("Konstanta", "nema");
			composit->InsertEndChild(operation);
		}
		if (typeid(*o) == typeid(CompositOperation)) {
			XMLElement* operation = xmlOperation->NewElement((((CompositOperation*)o)->getName()).c_str());
			operation->SetAttribute("Konstanta", "nema");
			composit->InsertEndChild(operation);
			exportComposit(i, (CompositOperation*)o);
		}
	}
	xmlOperation->SaveFile((comp->getName()).c_str());
}

File: POOPdz2/photoshop/src/projekat/XMLFormatter.java
package projekat;

import org.w3c.dom.*;
import org.xml.sax.SAXException;

import gui.Prozor;

import javax.xml.parsers.*;
import javax.xml.transform.OutputKeys;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerConfigurationException;
import javax.xml.transform.TransformerException;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;

import java.io.*;
import java.util.ArrayList;

public class XMLFormatter {
	public CompositOperation importComposit(Image im, String file) {
		CompositOperation temp = new CompositOperation(file);
		try {		
			DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
			DocumentBuilder dBuilder = factory.newDocumentBuilder();;
			Document doc = dBuilder.parse(file);
			doc.getDocumentElement().normalize();
			
			Element root = doc.getDocumentElement();
			NodeList children = root.getChildNodes();
			
			Node comp = children.item(1);
			NodeList fun = comp.getChildNodes();
			int i = 1;
			
			for(Node n = fun.item(i); n!=null; n = fun.item(i)) {
				NamedNodeMap att = n.getAttributes();
				if(n.getNodeName().equals("Add") || n.getNodeName().equals("Subtraction") || n.getNodeName().equals("InvertedSubtraction")
						|| n.getNodeName().equals("Multiplication") || n.getNodeName().equals("Divide") || n.getNodeName().equals("InvertedDivision")
						|| n.getNodeName().equals("Power") || n.getNodeName().equals("Log") || n.getNodeName().equals("Absolute")
						|| n.getNodeName().equals("Minimum") || n.getNodeName().equals("Maximum") || n.getNodeName().equals("Inversion") || n.getNodeName().equals("Grey") || n.getNodeName().equals("BlackAndWhite")
						|| n.getNodeName().equals("Median")) {
					String c = att.getNamedItem("Konstanta").getNodeValue();
					if(c.equals("nema")) {
						BasicOperation o = new BasicOperation(n.getNodeName(), 0);
						o.setNema(false);
						temp.addOperation(o);
					}
					else {
						BasicOperation o = new BasicOperation(n.getNodeName(), Integer.parseInt(c));
						o.setNema(true);
						temp.addOperation(o);
					}
					
				}
				else {				
					//if(im.getCompOperations().isEmpty()) {
					CompositOperation o = importComposit(im, n.getNodeName());
					//}
					//else {
					//	for(CompositOperation co : im.getCompOperations()) {
					//		String name = n.getNodeName();
					//		if(!co.getName().equals(name)) {
					//			CompositOperation o = importComposit(im, n.getNodeName());
						//		temp.addOperation(o);
						//		break;
						//	}
					//	}		
				//	}
				}
				i=i+2;
			}
			im.addComp(temp);
			return temp;
		} catch (ParserConfigurationException | SAXException | IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
			return null;
		}
	
	}

	public void exportComposit(CompositOperation c) {
		try {
			DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
			DocumentBuilder dBuilder = factory.newDocumentBuilder();
			Document doc = dBuilder.newDocument();

			Element root = doc.createElement("Composit");
			doc.appendChild(root);

			Element comp = doc.createElement(c.getName());
			root.appendChild(comp);

			for (Operation o : c.getComp()) {
				if (o instanceof BasicOperation) {
					Element oper = doc.createElement(((BasicOperation) o).getOperation());
					if (((BasicOperation) o).isNema())
						oper.setAttribute("Konstanta", "nema");
					else {
						oper.setAttribute("Konstanta", Integer.toString(((BasicOperation) o).getC()));
					}
					comp.appendChild(oper);
				} else {
					String s = (((CompositOperation) o).getName());
					Element oper = doc.createElement(s);
					oper.setAttribute("Konstanta", "nema");
					comp.appendChild(oper);
					exportComposit((CompositOperation) o);
				}
			}
			TransformerFactory transformerFactory = TransformerFactory.newInstance();
			Transformer transformer = transformerFactory.newTransformer();
			transformer.setOutputProperty("omit-xml-declaration", "yes");
			transformer.setOutputProperty(OutputKeys.INDENT, "yes");
			transformer.setOutputProperty(OutputKeys.METHOD, "xml");
			transformer.setOutputProperty("{http://xml.apache.org/xslt}indent-amount", "4");
			DOMSource domSource = new DOMSource(doc);
			StreamResult streamResult = new StreamResult(new File(c.getName()));
			transformer.transform(domSource, streamResult);

		} catch (ParserConfigurationException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (TransformerConfigurationException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (TransformerException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
	}

	public Image myImport(String file) {
		Image i = new Image();
		ArrayList<Layer> layList = new ArrayList<Layer>();
		try {
			DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
			DocumentBuilder dBuilder = factory.newDocumentBuilder();
			Document doc = dBuilder.parse(file);
			doc.getDocumentElement().normalize();
			Element root = doc.getDocumentElement();
			NodeList children = root.getChildNodes();

			Node dimensions = children.item(1);
			NodeList dim = dimensions.getChildNodes();
			Node w = dim.item(1);
			Node h = dim.item(3);
			int width = Integer.parseInt(w.getAttributes().getNamedItem("width").getNodeValue());
			int height = Integer.parseInt(h.getAttributes().getNamedItem("height").getNodeValue());

			Node layer = children.item(3);
			NodeList layers = layer.getChildNodes();
			int j = 1;
			for (Node n = layers.item(j); n != null; n = layers.item(j)) {
				NamedNodeMap atributes = n.getAttributes();
				String layName = atributes.getNamedItem("name").getNodeValue();
				int transparency = Integer.parseInt(atributes.getNamedItem("transparency").getNodeValue());
				boolean active = Boolean.parseBoolean(atributes.getNamedItem("active").getNodeValue());
				int selected = Integer.parseInt(atributes.getNamedItem("selected").getNodeValue());
				boolean visible = Boolean.parseBoolean(atributes.getNamedItem("visible").getNodeValue());

				BMPFormatter f = new BMPFormatter();
				Layer l = f.myImport(layName);
				if(!Prozor.chLayers.isEmpty())
					l.setChName(Prozor.chLayers.get(layName).getLabel());
				l.setName(layName);
				l.setActive(active);
				for (int jj = 0; jj < selected; jj++)
					;
				l.setSelected();
				l.setTransparency(transparency);
				l.setVisible(visible);
				layList.add(l);
				j = j + 2;
			}
			for (Layer l : layList) {
				i.addLayer(l);
			}

			j = 1;
			Node selection = children.item(5);
			NodeList selections = selection.getChildNodes();
			for (Node n = selections.item(j); n != null; n = selections.item(j)) {
				NamedNodeMap atributes = n.getAttributes();
				String name = atributes.getNamedItem("name").getNodeValue();
				boolean active = Boolean.parseBoolean(atributes.getNamedItem("active").getNodeValue());
				NodeList rectangles = n.getChildNodes();
				int k = 1;
				ArrayList<Rectangle> recList = new ArrayList<Rectangle>();
				for (Node m = rectangles.item(k); m != null; m = rectangles.item(k)) {
					NamedNodeMap atr = m.getAttributes();
					int y = Integer.parseInt(atr.getNamedItem("y").getNodeValue());
					int x = Integer.parseInt(atr.getNamedItem("x").getNodeValue());
					int heightR = Integer.parseInt(atr.getNamedItem("height").getNodeValue());
					int widthR = Integer.parseInt(atr.getNamedItem("width").getNodeValue());
					recList.add(new Rectangle(x, y, widthR, heightR));
					k = k + 2;
				}
				i.addSelection(name, new Selection(recList, name, active));
				j = j + 2;
			}
		//	Node comp = children.item(7);
		//	NodeList comps = comp.getChildNodes();
			//j=1;
			//for(Node n = comps.item(j); n!=null; n=comps.item(j)) {
			//	importComposit(i, n.getAttributes().getNamedItem("name").getNodeValue());
			//	j= j+2;
			//}

		} catch (ParserConfigurationException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (SAXException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		return i;

	}

	public void myExport(String file, Image i) {
		BMPFormatter bmp = new BMPFormatter();
		PAMFormatter pam = new PAMFormatter();
		try {
			DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
			DocumentBuilder dBuilder = factory.newDocumentBuilder();
			Document doc = dBuilder.newDocument();
			Element root = doc.createElement("Image");
			doc.appendChild(root);

			Element dimensions = doc.createElement("Dimensions");
			root.appendChild(dimensions);

			Element width = doc.createElement("Width");
			Element height = doc.createElement("Height");
			width.setAttribute("width", Integer.toString(i.getWidth()));
			height.setAttribute("height", Integer.toString(i.getHeight()));
			dimensions.appendChild(width);
			dimensions.appendChild(height);

			Element layers = doc.createElement("Layers");
			root.appendChild(layers);
			Image temp = new Image();
			for (int j = 0; j < i.getImage().size(); j++) {
				temp.addLayer(i.getImage().get(j));
				//bmp.myExport(i.getImage().get(j).getName() + ".bmp", temp);
				temp.getImage().remove(0);
				Element layer = doc.createElement("Layer");
				layer.setAttribute("transparency", Integer.toString(i.getImage().get(j).getTransparency()));
				layer.setAttribute("name", i.getImage().get(j).getName());
				layer.setAttribute("active", Boolean.toString(i.getImage().get(j).isActive()));
				layer.setAttribute("selected", Integer.toString(i.getImage().get(j).getSelected()));
				layer.setAttribute("visible", Boolean.toString(i.getImage().get(j).isVisible()));
				layers.appendChild(layer);
			}

			Element selections = doc.createElement("Selections");
			root.appendChild(selections);
			for (String s : i.getSelection_map().keySet()) {
				Element selection = doc.createElement("Selection");
				selections.appendChild(selection);
				selection.setAttribute("name", s);
				selection.setAttribute("active", Boolean.toString(i.getSelection_map().get(s).isActive()));
				for (Rectangle r : i.getSelection_map().get(s).getSelection()) {
					Element rectangle = doc.createElement("Rectangle");
					selection.appendChild(rectangle);
					rectangle.setAttribute("y", Integer.toString(r.getY()));
					rectangle.setAttribute("x", Integer.toString(r.getX()));
					rectangle.setAttribute("height", Integer.toString(r.getHeight()));
					rectangle.setAttribute("width", Integer.toString(r.getWidth()));
				}

			}

			TransformerFactory transformerFactory = TransformerFactory.newInstance();
			Transformer transformer = transformerFactory.newTransformer();
			transformer.setOutputProperty("omit-xml-declaration", "yes");
			transformer.setOutputProperty(OutputKeys.INDENT, "yes");
			transformer.setOutputProperty(OutputKeys.METHOD, "xml");
			transformer.setOutputProperty("{http://xml.apache.org/xslt}indent-amount", "4");
			DOMSource domSource = new DOMSource(doc);
			StreamResult streamResult = new StreamResult(new File(file));
			transformer.transform(domSource, streamResult);
		} catch (ParserConfigurationException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (TransformerConfigurationException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (TransformerException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
	}
}
